File: src/llm/claudeCli.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";

import { logClaudeCall, newId } from "../log/audit";

const execFileAsync = promisify(execFile);

// Headless driver for the already-authenticated local `claude` CLI (v2.x), used in
// place of an SDK — this box has no Anthropic API key, only the logged-in CLI.
// `--output-format json` wraps the run in an envelope whose `result` field is the
// assistant's final text; we return it RAW and let callers parse/validate it. Fails
// loud (throws) on a non-zero exit, a timeout, unparseable stdout, or an empty result.
const CLAUDE_BIN = "claude";
const MODEL = "claude-opus-4-8";
const TIMEOUT_MS = 120_000;
const MAX_BUFFER_BYTES = 32 * 1024 * 1024;

// Shape of the execFile rejection we read when the CLI fails (non-zero exit or timeout).
type ExecError = { code?: number | string; signal?: NodeJS.Signals | null; killed?: boolean; stderr?: string };

export async function runClaude(
  prompt: string,
  opts: { imagePaths?: string[]; caller?: string } = {},
): Promise<string> {
  const imagePaths = opts.imagePaths ?? [];
  const caller = opts.caller ?? "claude";
  const enableImages = imagePaths.length > 0;

  // With images we let Claude Code read them off disk (Read tool) and point it at the
  // absolute paths in the prompt; otherwise no tools are allowed at all.
  const fullPrompt = enableImages
    ? `${prompt}\n\nRead these image file(s) as part of this task:\n${imagePaths.join("\n")}`
    : prompt;
  const allowedTools = enableImages ? "Read" : "";

  const args = [
    "-p", fullPrompt,
    "--model", MODEL,
    "--output-format", "json",
    "--allowedTools", allowedTools,
    "--permission-mode", "bypassPermissions",
  ];

  // Audit every call: mint an id and start the wall clock before spawning so both the
  // success and failure paths log a complete record (best-effort; see ../log/audit).
  const callId = newId();
  const startedAt = Date.now();

  try {
    let stdout: string;
    try {
      ({ stdout } = await execFileAsync(CLAUDE_BIN, args, {
        timeout: TIMEOUT_MS,
        maxBuffer: MAX_BUFFER_BYTES,
        encoding: "utf8",
      }));
    } catch (cause) {
      const err = cause as ExecError;
      const reason = err.killed
        ? `timed out after ${TIMEOUT_MS}ms (signal ${err.signal})`
        : `exited with code ${err.code}`;
      const stderr = err.stderr ? `; stderr: ${err.stderr.trim()}` : "";
      throw new Error(`runClaude: \`${CLAUDE_BIN}\` ${reason}${stderr}`, { cause });
    }

    let envelope: unknown;
    try {
      envelope = JSON.parse(stdout);
    } catch (cause) {
      throw new Error(`runClaude: stdout was not valid JSON: ${stdout.slice(0, 1000)}`, { cause });
    }
    if (typeof envelope !== "object" || envelope === null) {
      throw new Error(`runClaude: expected a JSON object from --output-format json, got ${typeof envelope}`);
    }

    const env = envelope as Record<string, unknown>;
    const result = env.result;
    if (typeof result !== "string" || result.trim().length === 0) {
      throw new Error("runClaude: envelope.result is missing or empty");
    }

    // Capture the full envelope verbatim plus call metadata + the convenience fields the
    // audit summary reads (session_id / total_cost_usd / usage present when the CLI emits them).
    logClaudeCall({
      ...env,
      callId,
      caller,
      model: MODEL,
      prompt: fullPrompt,
      imagePaths,
      ok: true,
      durationMs: Date.now() - startedAt,
      result,
      session_id: env.session_id,
      total_cost_usd: env.total_cost_usd,
      duration_ms: env.duration_ms,
      duration_api_ms: env.duration_api_ms,
      num_turns: env.num_turns,
      usage: env.usage,
    });
    return result;
  } catch (err) {
    // Every failure path (exec error, unparseable stdout, non-object, empty result) is
    // audited before it propagates — then rethrow exactly as before (stay fail-loud).
    logClaudeCall({
      callId,
      caller,
      model: MODEL,
      prompt: fullPrompt,
      imagePaths,
      ok: false,
      durationMs: Date.now() - startedAt,
      error: err instanceof Error ? err.message : String(err),
    });
    throw err;
  }
}

File: backend/src/llm/claude_cli.py
"""Headless driver for the already-authenticated local `claude` CLI.

Mirrors src/llm/claudeCli.ts. Spawns `claude -p <prompt> --model claude-opus-4-8
--output-format json --permission-mode bypassPermissions`, parses the JSON
envelope, and returns envelope['result']. Fails loud on any error.
"""

import json
import subprocess

CLAUDE_BIN = "claude"
MODEL = "claude-opus-4-8"
TIMEOUT_SECONDS = 120
MAX_BUFFER_BYTES = 32 * 1024 * 1024


def run_claude(
    prompt: str,
    *,
    image_paths: list[str] | None = None,
    caller: str = "claude",
) -> str:
    """Run the local claude CLI and return the assistant's final text.

    Args:
        prompt: The prompt to send to Claude.
        image_paths: Optional list of image file paths for vision inputs.
        caller: Logical caller name for logging/auditing.

    Returns:
        The contents of envelope['result'] from the CLI JSON output.

    Raises:
        RuntimeError or subprocess.SubprocessError on failure.
    """
    image_paths = image_paths or []
    enable_images = len(image_paths) > 0

    full_prompt = prompt
    if enable_images:
        full_prompt = (
            f"{prompt}\n\nRead these image file(s) as part of this task:\n"
            + "\n".join(image_paths)
        )
    allowed_tools = "Read" if enable_images else ""

    args = [
        "-p",
        full_prompt,
        "--model",
        MODEL,
        "--output-format",
        "json",
        "--allowedTools",
        allowed_tools,
        "--permission-mode",
        "bypassPermissions",
    ]

    # Caller is accepted for API parity but the Python port does not yet audit.
    _ = caller

    try:
        result = subprocess.run(
            [CLAUDE_BIN, *args],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_SECONDS,
        )
    except subprocess.TimeoutExpired as exc:
        raise RuntimeError(
            f"run_claude: `{CLAUDE_BIN}` timed out after {TIMEOUT_SECONDS}s"
        ) from exc
    except subprocess.SubprocessError:
        raise

    if result.returncode != 0:
        stderr = f"; stderr: {result.stderr.strip()}" if result.stderr else ""
        raise RuntimeError(
            f"run_claude: `{CLAUDE_BIN}` exited with code {result.returncode}{stderr}"
        )

    try:
        envelope = json.loads(result.stdout)
    except json.JSONDecodeError as exc:
        raise RuntimeError(
            f"run_claude: stdout was not valid JSON: {result.stdout[:1000]}"
        ) from exc

    if not isinstance(envelope, dict):
        raise RuntimeError(
            f"run_claude: expected a JSON object from --output-format json, "
            f"got {type(envelope).__name__}"
        )

    output = envelope.get("result")
    if not isinstance(output, str) or output.strip() == "":
        raise RuntimeError("run_claude: envelope.result is missing or empty")

    return output
